text/0812-tracked-built-ins.md

```diff
@@ -0,0 +1,287 @@
+tracked version of JavaScript [Map][global-objects-map].
+
+**Note:** In this and the following examples, I am using `<template>` from
+[RFC #779][rfc-0779].
+
+```js
+// app/components/scoreboard.js
+import Component from '@glimmer/component';
+import { action } from '@ember/object';
+import { TrackedMap } from 'tracked-built-ins';
+
+const PLAYERS = [
+  { name: 'Zoey' },
+  { name: 'Tomster' },
+];
+
+export default class Scoreboard extends Component {
+  // Create a map from player -> score, with each player's score starting at 0
+  scores = new TrackedMap(PLAYERS.map(p => [p, 0]));
+
+  @action
+  incrementScore(player) {
+    let currentScore = this.scores.get(player);
+
+    this.scores.set(player, currentScore + 1);
+  }
+
+  <template>
+    {{#each-in this.scores |player score|}}
+      <div>
+        {{player.name}}: {{score}}
+    
+        <button type="button" {{on "click" (fn this.incrementScore player)}}>
+          +
+        </button>
+      </div>
+    {{/each-in}}
+  </template>
+}
+```
+
+Lets lake a look into another example involving a shopping cart, where we keep track
+of the items added to cart using tracked version of JavaScript's [Array][global-objects-array].
+
+```js
+// app/components/scoreboard.js
+import Component from '@glimmer/component';
+import { action } from '@ember/object';
+import { TrackedArray } from 'tracked-built-ins';
+
+export default class Cart extends Component {
+  items = new TrackedArray([
+    { productName: 'Jeans' },
+    { productName: 'T-shirt' },
+  ]);
+
+  @action
+  addPants(item) {
+    this.items.push(item);
+  }
+
+  <template>
+    {{#each this.items as |item|}}
+      <div>
+        {{item.productName}}
+      </div>
+    {{/each}}
+  
+    <button type="button" {{on "click" (fn this.addPants @pants)}}>
+      Add Pants
+    </button>
+  </template>
+}
+```
+
+## Drawbacks
+
+> Why should we *not* do this? Please consider the impact on teaching Ember,
+on the integration of this feature with other existing and planned features,
+on the impact of the API churn on existing apps, etc.
+
+> There are tradeoffs to choosing any path, please attempt to identify them here.
+
+## Alternatives
+
+- `tracked-built-ins` can be added only to the `app` blueprint and not to the `addon` blueprint.
+
+- Do nothing and recommend using re-setting the property to track data structure changes.
+
+  The tradeoffs of this approach described in "Motivation" section of this RFC.
+
+- Don't add `tracked-built-ins` to the blueprint and only update Ember Guides.
+
+  Ember.js is advertised as "batteries included" framework with complete out-of-the-box experience.
+  Asking users to install additional dependencies to achieve relatively common problem contradicts
+  with that paradigm.
+
+## Unresolved questions
+
+- Should we introduce these at something like `@glimmer/tracking` instead?
+
+  ```js
+  import { tracked, TrackedMap, TrackedSet /* etc. */ } from '@glimmer/tracking';
```

Comment:
Maybe a stupid question, but do we need the `TrackedMap`, `TrackedSet`, ... exports if we can do `tracked(Map)`, `tracked(Set)`, ...? To me, it seems nice to only have `import { tracked } from '@glimmer/tracking';`.

Reply:
Hum, that fells less intuitive though. In practice I've always used the imports instead of wrapping.

Reply:
the wrapping implies -- "well, what else can I wrap"?

Reply:
This is really important point which I forgot to ask @SergeAstapov to address in drafting, @bertdeblock – thank you for raising it! I’ll make sure we cover it in our discussion (scheduled for Friday, but we’ll see if we get to it this week – I’m going to try!).

For my part (we haven't talked about it at Framework yet at all), this is actually a piece I *actively dislike* about the current thing and think we should remove. It makes `tracked` do two *very* different things, and that increases its complexity in non-trivial ways:

- **In terms of the “reasoning footprint” for what `tracked` is and does as a consumer:** from a pedagogical perspective, we would need to cover how it both works as a decorator *and* as a wrapper which produces non-decorated types, and indeed we would need to cover what this means (and why the decorator doesn't do *more* magic, which is a thing which comes up regularly already):

   ```js
   import { tracked } from '@glimmer/tracking';
   
   class Example {
      @tracked doublyTracked = tracked([]);
   }
   ```

- **In terms of the internal implementation details:** [supporting all of these different flows][impl] is not terrible, but it’s not trivial, and it also gets to just delegate to [the base decorator][base], so it’s more complicated than that shows.[^perf]

- **In terms of making sure the types do the right things:** On the types front, consider: if you want `tracked<T>(map: Map<T>): TrackedMap<T>` and `tracked<T extends object>(obj: T): TrackedObject<T>` and `tracked<T>(array: Array<T>): TrackedArray<T>` and so on to work, we have to provide overloads for all of them, and TS’s overload handling is order-sensitive, so we have to make sure that we don't mess *that* up, etc. ([Here][types] is how that is done today.)

All of that is *doable*; there are no critical blockers here. It’s more that it is complexity and overhead on all fronts, which needs to be well-motivated… and it’s not clear to *me* what the benefit is over importing and using the type you want directly—especially because all the types have constructors which allow them to be constructed from their “vanilla” versions.

Net, I think it's actually much better to be clear about what we’re actually doing.

(Again, that’s just my take; we’ll see what everyone else says when we discuss it!)

[impl]: https://github.com/tracked-tools/tracked-built-ins/blob/9111be44d781448993c170fb68b59fb8a69667c2/addon/-private/decorator.ts#L42-L86
[base]: https://github.com/glimmerjs/glimmer.js/blob/2696e52bae1ddaad160fc61cc1466fc47466cafc/packages/%40glimmer/tracking/src/tracked.ts
[types]: https://github.com/tracked-tools/tracked-built-ins/blob/9111be44d781448993c170fb68b59fb8a69667c2/addon/-private/decorator.ts#L13-L35

[^perf]: I don’t expect it would make a *huge* difference, but given how commonly `@tracked` is used and the fact that setting it up is part of booting the app, I would personally prefer to keep this stuff out of the default flow for performance-reasons, too.

Reply:
Could also see it as a decorator

```ts
@tracked(Set)
ids: Set<number>
```

Reply:
Given that `tracked` has to operate differently for each type of thing I don't think it is a great idea to overload its signature.

Reply:
Yeah, probably right